src/core/utils/flash.cpp

```diff
@@ -39,44 +39,281 @@
+        }
+
+        otPlatFlashRead(aInstance, aSwapIndex, aOffset + sizeof(*this), aValue, aLength);
+        return aLength;
+    }
+
+    uint16_t GetKey(void) const { return mKey; }
+
+    uint16_t GetLength(void) const { return mLength; }
+
+    uint16_t GetDataSize(void) const
+    {
+        uint16_t wordMask = GetAlignmentMask();
+        return (mLength + wordMask) & ~wordMask;
+    }
+
+    static uint16_t GetAlignmentMask(uint8_t aFormat)
+    {
+        return (aFormat == kFormatV1) ? (kFlashWordSizeV1 - 1) : (kFlashWordSizeV2 - 1);
+    }
+
+    uint16_t GetSize(void) const { return sizeof(*this) + GetDataSize(); }
+
+    bool IsValid(void) const { return ((mFlags & (kFlagAddComplete | kFlagDelete)) == kFlagDelete); }
+
+    void SetDeleted(void)
+    {
+        memset(this, 0xff, sizeof(*this));
+        mFlags &= ~kFlagDelete;
+    }
+
+    bool IsFirst(void) const { return (mFlags & kFlagFirst) == 0; }
+
+    bool IsFree(void) const { return !IsAddBeginSet(); }
+
+protected:
+    bool IsIntegrityOk(void) const
+    {
+        OT_ASSERT(IsAddBeginSet());
+        return IsAddCompleteSet();
+        // TODO: validate GetFormat() == 0 && mReserved == 0xffff;
+    }
+
+    uint8_t GetFormat() const { return kFormatTop - (mFlags >> kFormatShift); }
+
+private:
+    uint16_t GetAlignmentMask(void) const { return GetAlignmentMask(GetFormat()); }
+
+    bool IsAddBeginSet(void) const { return (mFlags & kFlagAddBegin) == 0; }
+    bool IsAddCompleteSet(void) const { return (mFlags & kFlagAddComplete) == 0; }
+
+    enum
+    {
+        kFlagAddBegin    = 1 << 0, ///< 0 indicates record write has started, 1 otherwise.
+        kFlagAddComplete = 1 << 1, ///< 0 indicates record write has completed, 1 otherwise.
+        kFlagDelete      = 1 << 2, ///< 0 indicates record was deleted, 1 otherwise.
+        kFlagFirst       = 1 << 3, ///< 0 indicates first record for key, 1 otherwise.
+
+        /* The record's format version is in the upper nibble of mFlags */
+        kFormatTop   = 0xf,
+        kFormatShift = 12,
+
+        /* Initial value of flags: kFlagAddBegin, kFlagAddComplete and format version */
+        kFlagsInit = ((kFormatTop - kFormatV2) << kFormatShift) | (0xfff & ~(kFlagAddBegin | kFlagAddComplete)),
+    };
+
+    uint16_t mKey;
+    uint16_t mFlags;
+    uint16_t mLength;
+    uint16_t mReserved;
+} OT_TOOL_PACKED_END;
+static_assert(sizeof(RecordHeader) % kFlashWordSizeV2 == 0, "wrong RecordHeader size");
+
+OT_TOOL_PACKED_BEGIN
+class Record : public RecordHeader
+{
+public:
+    void Init(uint16_t aKey, const uint8_t *aData, uint16_t aDataLength)
+    {
+        OT_ASSERT(aDataLength <= kMaxDataSize);
+        RecordHeader::Init(aKey, aDataLength);
+
+        memcpy(mData, aData, aDataLength);
+
+        /* fill slack space with 0xff to avoid writing 0 more than once to each bit */
+        memset(mData + aDataLength, 0xff, GetDataSize() - aDataLength);
+
+        // TODO: calculate CRC
```

Comment:
What is the current thinking on where the CRC value will go? In the record header `mReserved` field?

Reply:
Yes, that's the idea.

Reply:
Done.